File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/59fd7fdca5d443cabf6a19325e6d8308.py
from collections import defaultdict, deque

maxn = 200007
mod = 998244353

n, t = 0, 0
c = [0] * maxn
s = [set() for _ in range(maxn)]
st = [set() for _ in range(maxn)]
deg = [set() for _ in range(3)]
degt = [set() for _ in range(3)]

def remove_leaf(s, deg, v):
    if not s[v]:
        deg[c[v]].discard((0, v))
        return
    u = next(iter(s[v]))

    assert v
    deg[c[v]].discard((len(s[v]), v))
    deg[c[u]].discard((len(s[u]), u))
    s[u].discard(v)
    s[v].discard(u)
    deg[c[u]].add((len(s[u]), u))

def id5(s, deg):
    while True:
        if not deg[0]:
            break
        ret = min(deg[0])
        if ret[0] > 1:
            break
        remove_leaf(s, deg, ret[1])

def solve(s, deg, beg):
    res = 0
    id5(s, deg)
    while True:
        if not deg[0] and not deg[1] and not deg[2]:
            break
        if not deg[beg]:
            beg = 3 - beg
            continue
        res += 1
        while True:
            if not deg[beg]:
                break
            ret = min(deg[beg])
            if ret[0] > 1:
                break
            remove_leaf(s, deg, ret[1])
            id5(s, deg)
        beg = 3 - beg
    return res

def main():
    global n, t, c, s, st, deg, degt
    t = int(input())
    for _ in range(t):
        n = int(input())
        for i in range(1, n + 1):
            c[i] = int(input())
        for _ in range(n - 1):
            u, v = map(int, input().split())
            s[u].add(v)
            s[v].add(u)
            st[u].add(v)
            st[v].add(u)
        for i in range(1, n + 1):
            deg[c[i]].add((len(s[i]), i))
            degt[c[i]].add((len(s[i]), i))
        print(max(min(solve(s, deg, 1), solve(st, degt, 2)), 1))

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Java/59fd7fdca5d443cabf6a19325e6d8308.java
import java.util.*;

public class 59fd7fdca5d443cabf6a19325e6d8308xcodeeval_processed_4000.json {
    static final int maxn = 200007;
    static final int mod = 998244353;

    static int n, t;
    static int[] c = new int[maxn];
    static Set<Integer>[] s = new HashSet[maxn], st = new HashSet[maxn];
    static Set<Pair>[] deg = new HashSet[3], degt = new HashSet[3];

    static void removeLeaf(Set<Integer>[] s, Set<Pair>[] deg, int v) {
        if (s[v].isEmpty()) {
            deg[c[v]].remove(new Pair(0, v));
            return;
        }
        int u = s[v].iterator().next();

        assert v > 0;
        deg[c[v]].remove(new Pair(s[v].size(), v));
        deg[c[u]].remove(new Pair(s[u].size(), u));
        s[u].remove(v);
        s[v].remove(u);
        deg[c[u]].add(new Pair(s[u].size(), u));
    }

    static void id5(Set<Integer>[] s, Set<Pair>[] deg) {
        while (true) {
            if (deg[0].isEmpty()) break;
            Pair ret = deg[0].iterator().next();
            if (ret.first > 1) break;
            removeLeaf(s, deg, ret.second);
        }
    }

    static int solve(Set<Integer>[] s, Set<Pair>[] deg, int beg) {
        int res = 0;
        id5(s, deg);
        while (true) {
            if (deg[0].isEmpty() && deg[1].isEmpty() && deg[2].isEmpty()) break;
            if (deg[beg].isEmpty()) {
                beg = 3 - beg;
                continue;
            }
            res++;
            while (true) {
                if (deg[beg].isEmpty()) break;
                Pair ret = deg[beg].iterator().next();

                if (ret.first > 1) break;
                removeLeaf(s, deg, ret.second);
                id5(s, deg);
            }
            beg = 3 - beg;
        }
        return res;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        t = sc.nextInt();
        while (t-- > 0) {
            n = sc.nextInt();
            for (int i = 1; i <= n; i++) {
                c[i] = sc.nextInt();
                s[i] = new HashSet<>();
                st[i] = new HashSet<>();
                deg[c[i]] = new HashSet<>();
                degt[c[i]] = new HashSet<>();
            }
            for (int i = 1; i < n; i++) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                s[u].add(v);
                s[v].add(u);
                st[u].add(v);
                st[v].add(u);
            }
            for (int i = 1; i <= n; i++) {
                deg[c[i]].add(new Pair(s[i].size(), i));
                degt[c[i]].add(new Pair(s[i].size(), i));
            }
            System.out.print(Math.max(Math.min(solve(s, deg, 1), solve(st, degt, 2)), 1L) + " ");
        }
        sc.close();
    }

    static class Pair {
        int first, second;

        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }
}
